services/user-svc/internal/hwkc/client.go

```diff
@@ -0,0 +1,243 @@
+package hwkc
+
+import (
+	"bytes"
+	"common"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/coreos/go-oidc"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2/clientcredentials"
+	"hwutil"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+type Client struct {
+	adminApiBaseUrl *url.URL
+	realmBaseUrl    *url.URL
+	http            *http.Client
+}
+
+func NewClientFromEnv(ctx context.Context) (*Client, error) {
+	clientSecret := hwutil.MustGetEnv("OAUTH_KC_CLIENT_SECRET")
+	return NewClient(ctx, "helpwave", common.GetOAuthIssuerUrl(), common.GetOAuthClientId(), clientSecret)
+}
+
+func NewClient(ctx context.Context, realm, issuerUrl, clientId, clientSecret string) (*Client, error) {
+	provider, err := oidc.NewProvider(context.Background(), issuerUrl)
+	if err != nil {
+		return nil, fmt.Errorf("cannot lookup provider: %w", err)
+	}
+
+	config := clientcredentials.Config{
+		ClientID:     clientId,
+		ClientSecret: clientSecret,
+		TokenURL:     provider.Endpoint().TokenURL,
+	}
+
+	if _, err := config.Token(ctx); err != nil {
+		return nil, fmt.Errorf("testint token exchange failed: %w", err)
+	}
+
+	parsedIssuerUrl, err := url.Parse(issuerUrl)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse issuer url: %w", err)
+	}
+
+	adminApiBaseUrl, err := url.Parse(fmt.Sprintf("%s://%s/admin/realms/%s", parsedIssuerUrl.Scheme, parsedIssuerUrl.Host, realm))
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse newly created admin api base url from issuer url: %w", err)
+	}
+
+	realmBaseUrl, err := url.Parse(fmt.Sprintf("%s://%s/realms/%s", parsedIssuerUrl.Scheme, parsedIssuerUrl.Host, realm))
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse newly created admin api base url from issuer url: %w", err)
+	}
+
+	client := &Client{
+		adminApiBaseUrl: adminApiBaseUrl,
+		realmBaseUrl:    realmBaseUrl,
+		http:            config.Client(ctx),
+	}
+
+	return client, err
+}
+
+func (c *Client) validateResponse(res *http.Response) error {
```

Comment:
this function is poorly worded, there is no validation going on

Reply:
Yeah kind of validation but I get your point there